### DISCUSSION TITLE
How to specify Type of string parameter

### DISCUSSION THREAD

**Discussion by dirthsj** [linq2db/linq2db#3779] [Q&A]
I have a query (a quite simple StartsWith) that gets converted to something similar to the following sql:

```
DECLARE @nameFragment_1 UniVarChar(5) -- String
SET     @nameFragment_1 = 'P119%'

SELECT TOP 100
  [x].[name]
FROM [dbo].[name_tbl]
WHERE [x].[name] LIKE @nameFragment_1 ESCAPE '~'
ORDER BY [x].[name]
```

This query times out after a couple minutes. If I change the parameter type slightly

```
DECLARE @nameFragment_1 VarChar(5) -- String
SET     @nameFragment_1 = 'P119%'

SELECT TOP 100
  [x].[name]
FROM [dbo].[name_tbl]
WHERE [x].[name] LIKE @nameFragment_1 ESCAPE '~'
ORDER BY [x].[name]
```

It executes in 6 ms.

How can I get linq2db to generate the second query instead? Is that even possible?

**Comment by viceroypenguin** [comment]
In general, the easiest way is to update the `[Column]` attribute of the `Name` property so that it is of type `UniVarChar`. This will let the engine use the correct parameter type to SARG the query.

**Reply by dirthsj** [reply]
Unfortunately that is not an option. The schema for this database was frozen several years ago.

**Reply by viceroypenguin** [reply]
Query: Are you saying the database schema is frozen or the ORM representation of the DB is frozen? 

I'm not suggesting to change the type in the database itself. I'm suggesting to update the `[Column]` attribute on the property in the L2DB DTO for the table.

**Reply by dirthsj** [reply]
The database schema is frozen. The ORM representation of the DB is generated using the linq2db CLI. I will experiment with changing it.

**Reply by dirthsj** [reply]
Update: I tried setting DbType to `"UniVarChar(50)"` and `"varchar(50)"` in the `[Column]` attribute. Neither had any effect on the generated SQL.

**Reply by MaceWindu** [reply]
That's sounds like a bug in LIKE typing. Could you check wether it works with `DataType = DataType.VarChar`?

**Reply by MaceWindu** [reply]
I've created #3787 to track it. As workaround you can use `Sql.Convert` extension over parameter value to specify type cast

**Reply by dirthsj** [reply]
Explicitly setting `DataType = DataType.VarChar` in the Column attribute worked, it generated with

`DECLARE @nameFragment_1 VarChar(5) -- AnsiString`

Thanks!
